psi4/src/psi4/libmints/oeprop.cc

```diff
@@ -1784,6 +1791,597 @@ PopulationAnalysisCalc::compute_lowdin_charges(bool print_output) {
+    int running_points = 0;
+
+    // Coordinates, weights, and rho (molecular electron density) at each grid point
+    std::vector<double> x_points(total_points, 0.0);
+    std::vector<double> y_points(total_points, 0.0);
+    std::vector<double> z_points(total_points, 0.0);
+    std::vector<double> weights(total_points, 0.0);
+    std::vector<double> rho(total_points, 0.0);
+
+    for (int b = 0; b < blocks.size(); b++) {
+        std::shared_ptr<BlockOPoints> block = blocks[b];
+        SharedVector rho_block;
+        std::size_t num_points = block->npoints();
+
+        point_func->set_max_functions(block->local_nbf());
+        point_func->set_max_points(num_points);
+        point_func->compute_points(block);
+        rho_block = point_func->point_values()["RHO_A"];
+
+        if (!same_dens_) {
+            rho_block->add(point_func->point_values()["RHO_B"]);
+        }
+
+        double* x = block->x();
+        double* y = block->y();
+        double* z = block->z();
+        double* w = block->w();
+
+        for (int p = 0; p < num_points; p++) {
+            x_points[running_points + p] = x[p];
+            y_points[running_points + p] = y[p];
+            z_points[running_points + p] = z[p];
+            weights[running_points + p] = w[p];
+            rho[running_points + p] = rho_block->get(p);
+        }
+
+        running_points += num_points;
+    }
+
+
+    //Calculates Number of Electrons Based on the Density Obtained using Grid Points
+    double rho_num_electrons = 0.0;
+    for (int p = 0; p < total_points; p++) {
+        rho_num_electrons += weights[p] * rho[p];
+    }
+
+    //Checks to see if the number of electrons calculated using the grid densities matches actual number of electrons in molecule
+    double mol_charge = mol->molecular_charge();
+    double mol_nuc_charge = 0.0;
+
+    for (int a = 0; a < num_atoms; a++) {
+        mol_nuc_charge += mol->Z(a);
+    }
+    double mol_electron_count = mol_nuc_charge - mol_charge;
+
+    double electron_diff = rho_num_electrons - mol_electron_count;
+
+    if (print_output) outfile->Printf("  Electron Count from Grid (Expected Number): %8.5f (%8.5f)\n", rho_num_electrons, mol_electron_count);
+    if (print_output) outfile->Printf("  Difference: %8.5f\n\n", electron_diff);
+
+    if (fabs(electron_diff) > 0.001) {
+        throw PsiException("Number of electrons calculated using the grid (" + std::to_string(round(100000.0 * rho_num_electrons) / 100000.0) + ") does not match the number of electrons in molecule. Try increasing the number of radial or spherical points (mbis_radial_points and mbis_spherical_points options).", __FILE__, __LINE__);
+    }
+
+    // Distances and displacements between grid points and nuclei
+    std::vector<std::vector<double>> distances(num_atoms, std::vector<double>(total_points, 0.0));
+    std::vector<std::vector<std::vector<double>>> disps(num_atoms, std::vector<std::vector<double>>(total_points, std::vector<double>(3, 0.0)));
```

Comment:
Is this style of nesting `std:vector` here and elsewhere OK for a large `total_points` (e.g millions of grid points)?

Reply:
That's an interesting question. A nested `std::vector` does require more overhead than a flattened `std:vector`, but is a little more readable (to me at least). Since this routine shouldn't be rate-limiting, I'd lean towards leaving as is.

Reply:
My intention was to double check if a possibly large nested `std::vector` was thought about. While this part might not be rate-limiting other parts could. I don't have principle objections if you think it's OK.